account_invoice_merge/models/account_invoice.py

```diff
@@ -0,0 +1,218 @@
+        ]
+        for field in ['sale_line_ids']:
+            if field in self.env['account.invoice.line']._fields:
+                fields.append(field)
+        return fields
+
+    @api.model
+    def _get_first_invoice_fields(self, invoice):
+        return {
+            'origin': '%s' % (invoice.origin or '',),
+            'partner_id': invoice.partner_id.id,
+            'journal_id': invoice.journal_id.id,
+            'user_id': invoice.user_id.id,
+            'currency_id': invoice.currency_id.id,
+            'company_id': invoice.company_id.id,
+            'type': invoice.type,
+            'account_id': invoice.account_id.id,
+            'state': 'draft',
+            'reference': '%s' % (invoice.reference or '',),
+            'name': '%s' % (invoice.name or '',),
+            'fiscal_position_id': invoice.fiscal_position_id.id,
+            'payment_term_id': invoice.payment_term_id.id,
+            'invoice_line_ids': {},
+            'partner_bank_id': invoice.partner_bank_id.id,
+        }
+
+    @api.multi
+    def do_merge(self, keep_references=True, date_invoice=False,
+                 remove_empty_invoice_lines=True):
+        """
+        To merge similar type of account invoices.
+        Invoices will only be merged if:
+        * Account invoices are in draft
+        * Account invoices belong to the same partner
+        * Account invoices are have same company, partner, address, currency,
+          journal, currency, salesman, account, type
+        Lines will only be merged if:
+        * Invoice lines are exactly the same except for the quantity and unit
+
+         @param self: The object pointer.
+         @param keep_references: If True, keep reference of original invoices
+
+         @return: new account invoice id
+
+        """
+
+        def make_key(br, fields):
+            list_key = []
+            for field in fields:
+                field_val = getattr(br, field)
+                if field in ('product_id', 'account_id'):
+                    if not field_val:
+                        field_val = False
+                if (isinstance(field_val, browse_record) and
+                        field != 'invoice_line_tax_ids' and
+                        field != 'sale_line_ids'):
+                    field_val = field_val.id
+                elif isinstance(field_val, browse_null):
+                    field_val = False
+                elif (isinstance(field_val, list) or
+                        field == 'invoice_line_tax_ids' or
+                        field == 'sale_line_ids'):
+                    field_val = ((6, 0, tuple([v.id for v in field_val])),)
+                list_key.append((field, field_val))
+            list_key.sort()
+            return tuple(list_key)
+
+        # compute what the new invoices should contain
+        new_invoices = {}
+        draft_invoices = [invoice
+                          for invoice in self
+                          if invoice.state == 'draft']
```

Comment:
Perhaps you can include [this](https://github.com/OCA/account-invoicing/pull/732) here also.

Reply:
I propose to add get draft invoices & _dirty_check overridable in this PR 
https://github.com/sylvainvh/account-invoicing/pull/1/files